src/develop/develop.h

```diff
@@ -138,7 +147,9 @@ typedef struct dt_develop_t
   uint32_t average_delay;
   uint32_t preview_average_delay;
   uint32_t preview2_average_delay;
-  struct dt_iop_module_t *gui_module; // this module claims gui expose/event callbacks.
+  struct dt_iop_module_t *gui_module;  // this module claims gui expose/event callbacks.
+  struct dt_iop_module_t *crop_module; // set by dt_dev_pixelpipe_synch() if an enabled crop module is included in history
```

Comment:
Instead of crop_module here I would suggest to add a new flag `IOP_FLAGS_CROPPING` to all modules doing a crop. That is, the current `crop` module, old & deprecated `crop and rotate` and `perspective correction`. Module wanting to use the full RAW image should ask for no-crop during pixel pipe and so in `pixelpipe_hb` just skip all modules with the `IOP_FLAGS_CROPPING` flags.

This is a bit like what was done for `IOP_FLAGS_ALLOW_FAST_PIPE` if I remember correctly.

Reply:
I agree we should allow **all** cropping modules to visualize the area (if that is what you are hinting at)

But - at least that's how this pr currently works - the canvas visualizing effect is done in darkroom expose so after processing the pipeline. `crop_module` is a) set while checking pipe nodes and b) is used after pipe processing by darkroom.

Reply:
Also we have `IOP_FLAGS_GUIDES_SPECIAL_DRAW` already

Reply:
> But - at least that's how this pr currently works - the canvas visualizing effect is done in darkroom expose so after processing the pipeline. `crop_module` is a) set while checking pipe nodes and b) is used after pipe processing by darkroom.

Yes, but then the knowledge of who is doing crop is in one place and this is not good to me. I still prefer having each module having a way to express its constraint/behavior. This makes the iop more independent of the  way they are used (the pixelpipe).

Reply:
I could say, we also have that knowledge about exposing some modules special thing there already.

But, yes it's bad. Then we should do that specific thing in module gui_expose and that must know about the valid crop area.

Reply:
> Then we should do that specific thing in module gui_expose

Maybe a bit late, my idea was really to skip all modules marked as doing crop directly in pixelpipe_hb as if the module was not enable at all. But I see you point we still need to know the actually crop area to display it... That's not an easy one and I'm sure you'll come up with the best compromise.